node/impl/full/eth_trace.go

```diff
@@ -0,0 +1,339 @@
+package full
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"fmt"
+	"io"
+
+	"github.com/multiformats/go-multicodec"
+	cbg "github.com/whyrusleeping/cbor-gen"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/builtin"
+	"github.com/filecoin-project/go-state-types/exitcode"
+
+	builtinactors "github.com/filecoin-project/lotus/chain/actors/builtin"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/chain/types/ethtypes"
+)
+
+func decodePayload(payload []byte, codec uint64) (ethtypes.EthBytes, error) {
+	if len(payload) == 0 {
+		return ethtypes.EthBytes(nil), nil
+	}
+
+	switch multicodec.Code(codec) {
+	case multicodec.Identity:
+		return ethtypes.EthBytes(nil), nil
+	case multicodec.DagCbor, multicodec.Cbor:
+		buf, err := cbg.ReadByteArray(bytes.NewReader(payload), uint64(len(payload)))
+		if err != nil {
+			return nil, xerrors.Errorf("decodePayload: failed to decode cbor payload: %w", err)
+		}
+		return buf, nil
+	case multicodec.Raw:
+		return ethtypes.EthBytes(payload), nil
+	}
+
+	return nil, xerrors.Errorf("decodePayload: unsupported codec: %d", codec)
+}
+
+// buildTraces recursively builds the traces for a given ExecutionTrace by walking the subcalls
+func buildTraces(ctx context.Context, traces *[]*ethtypes.EthTrace, parent *ethtypes.EthTrace, addr []int, et types.ExecutionTrace, height int64, sa StateAPI) error {
+	// lookup the eth address from the from/to addresses. Note that this may fail but to support
+	// this we need to include the ActorID in the trace. For now, just log a warning and skip
+	// this trace.
+	//
+	// TODO: Add ActorID in trace, see https://github.com/filecoin-project/lotus/pull/11100#discussion_r1302442288
+	from, err := lookupEthAddress(ctx, et.Msg.From, sa)
+	if err != nil {
+		log.Warnf("buildTraces: failed to lookup from address %s: %v", et.Msg.From, err)
+		return nil
+	}
+	to, err := lookupEthAddress(ctx, et.Msg.To, sa)
+	if err != nil {
+		log.Warnf("buildTraces: failed to lookup to address %s: %w", et.Msg.To, err)
+		return nil
+	}
+
+	trace := &ethtypes.EthTrace{
+		Action: ethtypes.EthTraceAction{
+			From:  from.String(),
+			To:    to.String(),
+			Gas:   ethtypes.EthUint64(et.Msg.GasLimit),
+			Input: nil,
+			Value: ethtypes.EthBigInt(et.Msg.Value),
+
+			FilecoinFrom:    et.Msg.From,
+			FilecoinTo:      et.Msg.To,
+			FilecoinMethod:  et.Msg.Method,
+			FilecoinCodeCid: et.Msg.CodeCid,
+		},
+		Result: ethtypes.EthTraceResult{
+			GasUsed: ethtypes.EthUint64(et.SumGas().TotalGas),
+			Output:  nil,
+		},
+		Subtraces:    len(et.Subcalls),
```

Comment:
What about calls to special methods from the evm that we skip? I think we may need to set this _after_ the fact.

Reply:
Done!